plugins/genrulebob/genrule.go

```diff
@@ -0,0 +1,413 @@
+	implicitOuts android.WritablePaths
+	rspfile      android.WritablePath
+}
+
+func (m *genrulebob) buildInouts(ctx android.ModuleContext, args map[string]string) {
+	ruleparams := blueprint.RuleParams{
+		Command: m.Properties.Cmd,
+		Restat:  true,
+	}
+
+	if m.Properties.Depfile {
+		args["depfile"] = ""
+	}
+	args["headers_generated"] = ""
+	args["srcs_generated"] = ""
+
+	if m.Properties.Rsp_content != nil {
+		args["rspfile"] = ""
+		ruleparams.Rspfile = "${rspfile}"
+		ruleparams.RspfileContent = *m.Properties.Rsp_content
+	}
+
+	rule := ctx.Rule(pctx, "bob_gen_"+ctx.ModuleName(), ruleparams, utils.SortedKeys(args)...)
+
+	for _, sio := range m.inouts {
+		// `args` is slightly different for each inout, but blueprint's
+		// parseBuildParams() function makes a deep copy of the map, so
+		// we're OK to re-use it for each target.
+		if m.Properties.Depfile {
+			args["depfile"] = sio.depfile.String()
+		}
+		if m.Properties.Rsp_content != nil {
+			args["rspfile"] = sio.rspfile.String()
+		}
+		args["headers_generated"] = strings.Join(utils.Filter(utils.IsHeader, sio.out.Strings()), " ")
+		args["srcs_generated"] = strings.Join(utils.Filter(utils.IsNotHeader, sio.out.Strings()), " ")
+
+		ctx.Build(pctx,
+			android.BuildParams{
+				Rule:            rule,
+				Description:     "gen " + ctx.ModuleName(),
+				Inputs:          sio.in,
+				Implicits:       sio.implicitSrcs,
+				Outputs:         sio.out,
+				ImplicitOutputs: sio.implicitOuts,
+				Args:            args,
+				Depfile:         sio.depfile,
+			})
+	}
+}
+
+func (m *genrulebob) calcExportGenIncludeDirs(mctx android.ModuleContext) android.Paths {
+	var allIncludeDirs android.Paths
+
+	// Add our own include dirs
+	for _, dir := range m.Properties.Export_gen_include_dirs {
+		allIncludeDirs = append(allIncludeDirs, android.PathForModuleGen(mctx, dir))
+	}
+
+	// Add include dirs of our all dependencies
+	mctx.WalkDeps(func(child android.Module, parent android.Module) bool {
+		if mctx.OtherModuleDependencyTag(child) != encapsulatesTag {
+			return false
+		}
+		if cmod, ok := child.(genruleInterface); ok {
+			for _, dir := range cmod.GeneratedHeaderDirs() {
+				allIncludeDirs = append(allIncludeDirs, dir)
+			}
+		}
+		return true
+	})
+
+	// Make unique items as for recursive passes it may contain redundant ones
+	return android.FirstUniquePaths(allIncludeDirs)
+}
+
+func getDepfileName(s string) string {
+	return s + ".d"
+}
+
+func (m *genrulebob) inoutForSrc(ctx android.ModuleContext, re *regexp.Regexp, source android.Path) (sio soongInout) {
+	replaceSource := func(props []string) (new_props []string) {
+		for _, prop := range props {
+			new_props = append(new_props, re.ReplaceAllString(source.Rel(), prop))
+		}
+		return
+	}
+	mop := m.Properties.Multi_out_props
+
+	sio.in = android.Paths{source}
+	sio.out = pathsForModuleGen(ctx, replaceSource(mop.Replace))
+	sio.implicitSrcs = android.PathsForModuleSrc(ctx, replaceSource(mop.Implicit_srcs))
```

Comment:
I'd expect implicit sources derived from a specific source file to be relative to the same directory as `source`.
Mostly this will be the module's source dir - so `PathsForModuleSrc` would be correct. However, if source has come from another generated module, this will be somewhere in the build dir instead.

Can we rewrite this as something along the lines of `filepath.Join(source.NonRel(), replaceSource(mop.Implicit_srcs))`? Note that you might have to do something weird to get the non-rel component of `source`, as `paths.go` doesn't provide a built-in way to do this.

Reply:
I'm not sure `PathsForSource` will work either - it's closer, but won't allow us to use paths from the `out` directly:
```
// PathForSource joins the provided path components and validates that the result
// neither escapes the source dir nor is in the out dir.
// On error, it will return a usable, but invalid SourcePath, and report a ModuleError.
func PathForSource(ctx PathContext, pathComponents ...string) SourcePath {
```
See "nor is in the out dir"...

Maybe you could check explicitly for if `source` is an `android.OutputPath`, and when it is, use something like `OutputPath.WithoutRel().Join(...)`?

Reply:
Thanks!